### ISSUE TITLE
Give example applications of Willow

### ISSUE THREAD

**Issue by sgwilym** [earthstar-project/willowprotocol.org#61]
> > Still confused here. What is the actual, concrete ‘as-a-user-I-want-to’ application for which this is meant to be an ideal fit? Sorry if a dumb question.

> > Yeah, it would have helped me if they walked through what it actually means to "use" Willow. Do I install something like Dropbox on my computer? Do I write code that calls Willow as a library?

> > Same here. no clue what it does.. It could be a syncthing/dropbox something. It could be some sharing protocol. I dunno

> Probably warrants some candid words on the landing page.

**Comment by joakim** [comment]
Nice, that's definitely more clear. Looks great too!

> Only devices which have been given explicit consent to store data can receive it.

I was confused when I first read this, maybe "access" is a better word than "receive"?

It's all sort of abstract. I think what any protocol needs is a "killer app", basically Earthstar :) I think that's what the quotes were getting at, that without an actual implementation of the protocol it's hard to grasp what it is, how it can be used.

---

That was fun :) I'm proud to have contributed a _tiny_ bit to Willow. Interesting choice of epoch! Why not J2000? To allow backdating?

**Comment by joakim** [comment]
We're way off-topic now :) Is it OK to continue down the rabbit hole here?

For the record, I'm very much an amateur, not even an astronomer. I'm still learning as I go.

Looks like they use as [the basis for their calculations](https://docs.rs/hifitime/4.2.2/src/hifitime/timescale/mod.rs.html#139) NTP's epoch, which is `1900-01-01 00:00`. They then define [`J1900_REF_EPOCH`](https://docs.rs/hifitime/4.2.2/src/hifitime/timescale/mod.rs.html#25-31) as that + 12 hours, which should have been a subtraction according to [NASA's definition](https://naif.jpl.nasa.gov/pub/naif/toolkit_docs/FORTRAN/spicelib/j1900.html). Can't say if that's a bug or an intentional change to make J1900 occur exactly one century before J2000. The `JD_J1900` constant is correct (`2_415_020 = 1899-12-31 12:00:00`) though, so it looks like a bug to me. In any case, it's not standard.

(AFAIK, the reason J1900 is half a day before the century began is because `2415020` is the integer that contains the instant of the midnight, which is `2415020.5`. For some reason J2000 doesn't do that.)

`JD_J2000` is correct (`2_451_545 = 2000-01-01 12:00:00`) and `J2000_REF_EPOCH` looks right in code, but the documentation is wrong, it's not midnight but noon.

> The crate also has [this constant for the unix reference epoch](https://docs.rs/hifitime/4.2.2/src/hifitime/timescale/mod.rs.html#73-77). Does this match JD 2441317.5 accoding to your understanding? If so, then I think that'll simply become our recommended epoch.

No, that's the common Unix epoch `1970-01-01 00:00:00`. It's ubiquitous in computing, so it seems like an obvious choice, I have even suggested it before :) Now I see that there's room for confusion and trouble when using a 1970 epoch with TAI.

1. Risk of people thinking that Unix time is essentially TAI (surprisingly common)
2. Is it `1970-01-01 UTC` or `1970-01-01 TAI`? Libraries seem to do different things, as we have seen. The offset is 8s (and 82μs) at the Unix epoch, so important to get that right.
3. You'd still have to deal with the [complexity](https://github.com/qntm/t-a-i#:~:text=Prior%20to%201%20January%201972,%20the%20relationship%20was%20quite%20complex) when converting timestamps before 1972 (arguably not something one has to do very often, but if one claims to support something one should be able to do so, requiring complex calculations)

So in conclusion, I'm still partial to 1972 TAI (10s offset from UTC). It would stand apart from Unix time and avoid some of these issues. I don't think you even need a library, you only need access to a table of leap seconds (the OS has one), the rest is simple arithmetic as far as I can tell. It seems like the most elegant solution.

**Comment by AljoschaMeyer** [comment]
Thank you, that was *very* helpful! And yes, I am totally fine with going off-topic here =)

So... I think I will open an issue in the repo of that crate, for starters. And then I'll change the spec draft and implementation to using the 1972-01-01 00:00:00 UTC (JD 2441317.5) epoch you suggested.

As for updating our implementation: I'm still a bit lost on how to express that epoch in hifitime. Is [`from_jdt_et(2441317.5)`](https://docs.rs/hifitime/4.2.2/hifitime/struct.Epoch.html#method.from_jde_et) the right thing? I'm somewhat skeptical about using floats to express a fraction of a day which should then be converted to the exact right nanosecond =S

**Comment by joakim** [comment]
That's a fair analysis! I think it's wise to pick an epoch that's supported by libraries that will handle leap seconds for you. And I do see the advantage of using a standard epoch.

I'm not convinced that J1900 is the best choice though.

Some thoughts:

- It doesn't really make sense to talk about TAI timestamps before 1960, when TAI was formally defined. Before the atomic clocks started ticking, there were no TAI seconds. IAU deems TAI < 1960 invalid, see [this relevant note](https://docs.rs/hifitime/latest/hifitime/index.html#important) in hifitime.
- Pre-1972 epochs are not exactly frictionless for Unix timestamps < 1972, libraries differ a lot in how they convert these. Even NASA's SPICE and IAU/IERS disagree on what to do, hifitime getting [caught in the middle](https://docs.rs/hifitime/latest/src/hifitime/epoch/initializers.rs.html#205-210).

If representing older timestamps is not essential, I think J2000 would be a good choice. It avoids the tricky conversion issues and is the current standard epoch in astronomy, clearly [defined by international agreement](https://en.wikipedia.org/wiki/Epoch_(astronomy)#:~:text=The%20currently%20used%20standard%20epoch%20%22J2000%22%20is%20defined%20by%20international%20agreement%20to%20be%20equivalent%20to). (Note that it's not noon UTC, but noon TT.)

I checked IAU's [standard library](https://www.iausofa.org/2023-10-11c), and the function `epj` for converting Julian Date to the J2000 epoch has this comment:

> For historical reasons, the time scale formally associated with Julian Epoch is TDB (or TT, near enough).  However, Julian Epoch can be used more generally as a calendrical convention to represent other time scales _**such as TAI**_ and TCB.  This is analogous to Julian Date, which was originally defined specifically as a way of representing Universal Times but is now routinely used for any of the regular time scales.

---

> As for updating our implementation: I'm still a bit lost on how to express that epoch in hifitime. Is [`from_jdt_et(2441317.5)`](https://docs.rs/hifitime/4.2.2/hifitime/struct.Epoch.html#method.from_jde_et) the right thing? I'm somewhat skeptical about using floats to express a fraction of a day which should then be converted to the exact right nanosecond =S

For good reason :) `f64` has only 20 μs precision with JD.

JDE (Julian Day Edwards) is [an enterprisey format](https://docs.oracle.com/cd/E26228_01/doc.93/e21961/julian_date_conv.htm). Maybe just use [`from_gregorian_utc`](https://docs.rs/hifitime/4.2.2/hifitime/struct.Epoch.html#method.from_gregorian_utc) to be as clear as possible? Everybody understands UTC. With J2000, I think you could just do [`from_et_seconds(0)`](https://docs.rs/hifitime/latest/src/hifitime/epoch/initializers.rs.html#194-197).

**Comment by AljoschaMeyer** [comment]
Alright, then unless I wake up with new thoughts on this, it'll be J2000!
